### ISSUE TITLE
Poor performance when focus changes 

### ISSUE THREAD

**Issue by russelltg** [hyprwm/Hyprland#2501]
Setup: 

Arch Linux
hyprland-git `r2982.e1edfde5-1`
i9-11900H CPU/GPU
4k60 display

Steps to repro:
Enable the debug overlay
Open two windows
Switch focus

Result:
Watch FPS drop significantly for a few seconds

Expected result:
FPS maintains constant 60Hz

Additional info:
It seems to be bound on the GPU here, but I don't think it's maxxing out the GPU...looking at intel_gpu_top, when spamming focus changes, GPU usage is around 88%, but for some reason it's only running at 500 MHz, with a max stated as ~1500MHz--so it seems like the GPU isn't clocking up correctly or something like that?

When just sitting on a desktop running a Firefox instance, intel_gpu_top and htop in terminals, Hyprland uses ~38% CPU (htop numbers, so 100% means ONE core fully utilized). When spamming switching focus, it goes down to 10%, likely b/c waiting on the GPU. 

It seems like switching focus shouldn't be such a GPU intensive task, and it seems like my system has the headroom to work much nicer if the GPU were able to clock up properly. 

Given the GPU clocking stuff, this may be a driver bug of sorts, or at least partially?

In general performance on this 4k display is pretty poor with this GPU, and it's pretty smooth in Windows and it's very smooth running 1x 1440p60 + 1x 1440p144, which is more pixels per second, so something particular to the 4k case seems to be happening? 

Maybe related to https://github.com/hyprwm/Hyprland/issues/2305 also

**Comment by vaxerski** [comment]
sir gpus dont work like that

do you have any animations playing when you do that? what if you set `animations:enabled` to 0?

**Comment by russelltg** [comment]
> gpus dont work like that

Which part lol?

> what if you set animations:enabled to 0?

Same result

**Comment by romanstingler** [comment]
I have Arch with the latest git build from today. 
i7-10750H
also 2x4k display
and if I switch windows my intel_top shows me an increase in the GPU frequency. 

Can you check if your device is connected to AC and check the CPU governor or if you have any power saving settings/tools active.

**Comment by russelltg** [comment]
Yes, connected via AC and using the powersave governor (as suggested https://www.gfxstrand.net/faith/blog/2020/11/getting-the-most-out-of-your-intel/ -- apparently it gives more of the power budget to the GPU).

Here's a little screenrecording--performance is obviously impacted by screenrecording--seems to be ~10FPS difference.


https://github.com/hyprwm/Hyprland/assets/7105082/70f60a2e-fc4a-46bd-a4c2-7f58d634d1e1

It seems very strange to me that the performance is SO bad, and that the GPU is clocked down so much...

**Comment by romanstingler** [comment]
I can confirm, that I could replicate that, but in my case it was just dropping to 30fps for a short period of time and increased slowly to 60fps. 

When I tested setting the GPU frequency to max (for my iGPU) 
`echo 1150 | tee /sys/class/drm/card0/gt_min_freq_mhz` 
the FPS stayed constant at 60 there was no way it would drop. 

So my guess is that other projects are rendering more often or deliberately to keep the freq higher (it's a guess, haven't compared yet).

My nvtop result shows that there are frequency drops even if I continually switch focus
![image](https://github.com/hyprwm/Hyprland/assets/5576250/cd65b71b-35c9-496d-847b-fb5f719c1ea5)

**Comment by vaxerski** [comment]
does this happen on sway too? maybe its just the app redrawing itself or something?

**Comment by russelltg** [comment]
I collected a bit of data with a simple setup: 
 
1x4k60 display
 i9-11900H CPU/GPU

this on the display:
![image](https://github.com/hyprwm/Hyprland/assets/7105082/87025f2e-a3ff-4df7-99b6-0edaa5eb8df9)


| WM | freq setting | FPS | GPU Clock speed (MHz) | GPU Utilization |
|--------|--------|--------|--------|--------|
| sway | -d | 60 | 414 | 23% |
| sway | -m | 60 | 870 | 16% |
| Hyprland (overlay ON) | -d | 30* | 480 | 52% | 
| Hyprland (overlay ON) | -m | 60** | 1410 | 50% |
| Hyprland (overlay OFF) | -d  | 60** | 550 | 85% |
| Hyprland (overlay ON) | -m | 60 | 1205 | 38% |

I first did all the tests with the debug overlay on, then re-did them after I realized it makes a significant performance difference

`*`: oscillates between 30 and 60, spending probably 90% of time in 30FPS
`**`: misses exactly one frame per second (don't mind the bit at the end, that was from the screenshot tool)
![image](https://github.com/hyprwm/Hyprland/assets/7105082/205f137f-c0b4-4515-862d-ffebf2a79257)

**Comment by vaxerski** [comment]
I pushed some commits to master, can you check the window switching perf issue?

**Comment by romanstingler** [comment]
> I pushed some commits to master, can you check the window switching perf issue?

I rebuilt it with the current commit and tested it. 
As soon as wf-recorder is activated, it drops to 18 fps 

![image](https://github.com/hyprwm/Hyprland/assets/5576250/edea7875-8a82-4d2b-b7e2-6857ee251282)

**Comment by vaxerski** [comment]
thats a separate thing, we're talking about the "when focus changes" here.

**Comment by romanstingler** [comment]
> thats a separate thing, 
missing screencopy-dmabuf? 


> we're talking about the "when focus changes" here.
yeah, with disabled animations the FPS looks promising when switching between windows. 
Stays above 55 FPS no matter how fast or what windows I switch (before drops down to 30fps). 

@russelltg what is your experience??
@vaxerski from my side a big thanks to you

**Comment by PoSayDone** [comment]
Also maybe related to #2484, cause in my case gpu clock also not rising properly and stays on 300MHz and cause fps drops.

**Comment by vaxerski** [comment]
based on reports from @romanstingler I am closing this as complete.

Poor performance of screencopy can be opened as a separate issue.
